pkg/sync/syncer.go

```diff
@@ -928,6 +1038,191 @@ func (s *syncer) syncGrantsForResource(ctx context.Context, resourceID *v2.Resou
+
+	// Peek the next action on the stack
+	if len(graph.Actions) == 0 {
+		return true, nil
+	}
+	action := graph.Actions[0]
+
+	l = l.With(zap.String("source_entitlement_id", action.SourceEntitlementID), zap.String("descendant_entitlement_id", action.DescendantEntitlementID))
+
+	// Fetch source and descendant entitlement
+	sourceEntitlement, err := s.store.GetEntitlement(ctx, &reader_v2.EntitlementsReaderServiceGetEntitlementRequest{
+		EntitlementId: action.SourceEntitlementID,
+	})
+	if err != nil {
+		l.Error("runGrantExpandActions: error fetching source entitlement", zap.Error(err))
+		return false, fmt.Errorf("runGrantExpandActions: error fetching source entitlement: %w", err)
+	}
+
+	descendantEntitlement, err := s.store.GetEntitlement(ctx, &reader_v2.EntitlementsReaderServiceGetEntitlementRequest{
+		EntitlementId: action.DescendantEntitlementID,
+	})
+	if err != nil {
+		l.Error("runGrantExpandActions: error fetching descendant entitlement", zap.Error(err))
+		return false, fmt.Errorf("runGrantExpandActions: error fetching descendant entitlement: %w", err)
+	}
+
+	// Fetch a page of source grants
+	sourceGrants, err := s.store.ListGrantsForEntitlement(ctx, &reader_v2.GrantsReaderServiceListGrantsForEntitlementRequest{
+		Entitlement: sourceEntitlement.GetEntitlement(),
+		PageSize:    1000,
+		PageToken:   action.PageToken,
+	})
+	if err != nil {
+		l.Error("runGrantExpandActions: error fetching source grants", zap.Error(err))
+		return false, fmt.Errorf("runGrantExpandActions: error fetching source grants: %w", err)
+	}
+
+	for _, sourceGrant := range sourceGrants.List {
+		// Unroll all grants for the principal on the descendant entitlement. This should, on average, be... 1.
+		descendantGrants := make([]*v2.Grant, 0, 1)
+		pageToken := ""
+		for {
+			req := &reader_v2.GrantsReaderServiceListGrantsForEntitlementRequest{
+				Entitlement: descendantEntitlement.GetEntitlement(),
+				PrincipalId: sourceGrant.GetPrincipal().GetId(),
+				PageSize:    1000,
+				PageToken:   pageToken,
+				Annotations: nil,
+			}
+
+			resp, err := s.store.ListGrantsForEntitlement(ctx, req)
+			if err != nil {
+				l.Error("runGrantExpandActions: error fetching descendant grants", zap.Error(err))
+				return false, fmt.Errorf("runGrantExpandActions: error fetching descendant grants: %w", err)
+			}
+
+			descendantGrants = append(descendantGrants, resp.List...)
+			pageToken = resp.NextPageToken
+			if pageToken == "" {
+				break
+			}
+		}
+
+		// If we have no grants for the principal in the descendant entitlement, make one.
+		directGrant := true
+		if len(descendantGrants) == 0 {
+			directGrant = false
+			// TODO(morgabra): This is kinda gnarly, grant ID won't have any special meaning.
+			// FIXME(morgabra): We should probably conflict check with grant id?
+			descendantGrant := grant.NewGrant(descendantEntitlement.GetEntitlement().GetResource(), descendantEntitlement.GetEntitlement().DisplayName, sourceGrant.GetPrincipal().GetId())
```

Comment:
Is `DisplayName` here a typo?

Reply:
I understand what is going on now. I think in this case, I'm not sure we want to use the `NewGrant` helper here, as it essentially requires that the connector dev is using the helpers to create their resources.